packages/core/src/storages/request_list.ts

```diff
@@ -593,21 +592,31 @@ export class RequestList {
         if (uniqueKey) {
             this.reclaimed.delete(uniqueKey);
             const index = this.uniqueKeyToIndex[uniqueKey];
-            return this.requests[index];
+            return this.ensureRequest(this.requests[index], index);
         }
 
         // Otherwise return next request.
         if (this.nextIndex < this.requests.length) {
-            const request = this.requests[this.nextIndex];
-            this.inProgress.add(request.uniqueKey);
+            const index = this.nextIndex;
+            const request = this.requests[index];
+            this.inProgress.add(request.uniqueKey!);
             this.nextIndex++;
             this.isStatePersisted = false;
-            return request;
+            return this.ensureRequest(request, index);
         }
 
         return null;
     }
 
+    private ensureRequest(requestLike: Request | RequestOptions, index: number): Request {
+        if (requestLike instanceof Request) {
+            return requestLike;
+        }
+
+        this.requests[index] = new Request(requestLike);
+        return this.requests[index] as Request;
+    }
```

Comment:
This is a bit unclear to me, why do we want to store the `Request` object back in the array? 

Wouldn't something like 
```typescript
makeRequest(r: RequestLike) : Request {
   if(r instanceof Request) return r;
   return new Request(r);
}
``` 

work just fine? Or do the (possible) repeated request constructions pose a performance problem?

Reply:
because when we reclaim requests, we need to reclaim the same request object that was returned via `fetchNextRequest` (otherwise tests break, haven't been debugging why more in detail)

will add a comment for this

Reply:
`fetchNextRequest` would still return a `Request` instance... the same one we would pass to `reclaim` later even.

But whatever, let's treat this as a hotfix, I assume the Delivery team depends on this. We'll revisit this in v4 either way, judging by the comments in `Request`?

Reply:
the thing is, you need to persist the reclaimed requests, and that needs to happen with their metadata (e.g. `userData`), if we wouldn't do this, we would loose that completely on migration. at least i think so.

Reply:
I see, sounds legit 👍🏽